Cabal/src/Distribution/Simple/GHC/Build/Modules.hs

```diff
@@ -0,0 +1,349 @@
+
+  -- If there aren't modules, or if we're loading the modules in repl, don't build.
+  unless (forRepl || (null inputFiles && null inputModules)) $ liftIO $ do
+    -- See Note [Building Haskell Modules accounting for TH]
+    let
+      neededWays =
+        wantedWays
+          <> Set.fromList
+            -- ROMES:TODO: You also don't need this if you are using an external interpreter!!
+            [defaultGhcWay | doingTH && defaultGhcWay `Set.notMember` wantedWays]
+
+      -- If we need both static and dynamic, use dynamic-too instead of
+      -- compiling twice (if we support it)
+      useDynamicToo =
+        StaticWay `Set.member` neededWays
+          && DynWay `Set.member` neededWays
+          && supportsDynamicToo comp
+          && null (hcSharedOptions GHC bi)
+
+      -- The ways we'll build, in order
+      orderedBuilds
+        -- If we can use dynamic-too, do it first. The default GHC way can only
+        -- be static or dynamic, so, if we build both right away, any modules
+        -- possibly needed by TH later (e.g. if building profiled) are already built.
+        | useDynamicToo =
+            [buildStaticAndDynamicToo]
+              ++ (runGhcProg . buildOpts <$> Set.toList neededWays \\ [StaticWay, DynWay])
+        -- Otherwise, we need to ensure the defaultGhcWay is built first
+        | otherwise =
+            runGhcProg . buildOpts <$> sortOn (\w -> if w == defaultGhcWay then 0 else fromEnum w + 1) (Set.toList neededWays)
+
+      buildStaticAndDynamicToo = do
+        runGhcProg dynTooOpts
+        case (hpcdir Hpc.Dyn, hpcdir Hpc.Vanilla) of
+          (Flag dynDir, Flag vanillaDir) ->
+            -- When the vanilla and shared library builds are done
+            -- in one pass, only one set of HPC module interfaces
+            -- are generated. This set should suffice for both
+            -- static and dynamically linked executables. We copy
+            -- the modules interfaces so they are available under
+            -- both ways.
+            copyDirectoryRecursive verbosity dynDir vanillaDir
+          _ -> return ()
+     in
+      -- REVIEW:ADD? info verbosity "Building Haskell Sources..."
+      sequence_ orderedBuilds
+  return buildOpts
+
+data BuildWay = StaticWay | DynWay | ProfWay
+  deriving (Eq, Ord, Show, Enum)
+
+-- | Returns the object/interface extension prefix for the given build way (e.g. "dyn_" for 'DynWay')
+buildWayPrefix :: BuildWay -> String
+buildWayPrefix = \case
+  StaticWay -> ""
+  ProfWay -> "p_"
+  DynWay -> "dyn_"
+
+-- | Returns the corresponding 'Hpc.Way' for a 'BuildWay'
+buildWayHpcWay :: BuildWay -> Hpc.Way
+buildWayHpcWay = \case
+  StaticWay -> Hpc.Vanilla
+  ProfWay -> Hpc.Prof
+  DynWay -> Hpc.Dyn
+
+-- | Returns a function to extract the extra haskell compiler options from a
+-- 'BuildInfo' and 'CompilerFlavor'
+buildWayExtraHcOptions :: BuildWay -> Maybe (CompilerFlavor -> BuildInfo -> [String])
```

Comment:
Why the `Maybe`? Is there a case/situation where this function should return `Nothing`?

Reply:
Not anymore!